Add ERC20 and ERC777 fixed supply presets #2377

ashwinYardi opened the pull request:
<!-- 0. 🎉 Thank you for submitting a PR! -->

<!-- 1. Does this close any open issues? Please list them below. -->
Open issue: https://github.com/OpenZeppelin/openzeppelin-contracts/issues/2377

<!-- Keep in mind that new features have a better chance of being merged fast if
they were first discussed and designed with the maintainers. If there is no
corresponding issue, please consider opening one for discussion first! -->

Additions #

1. Added ERC20 preset contract with with fixed initial supply.
2. Added ERC777 preset contract with fixed initial supply.
3. Added tests for both the presets.

<!-- 2. Describe the changes introduced in this pull request. -->
<!--    Include any context necessary for understanding the PR's purpose. -->

<!-- 3. Before submitting, please make sure that you have:
  - reviewed the OpenZeppelin Contributor Guidelines
    (https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/CONTRIBUTING.md),
  - added tests where applicable to test new functionality,
  - made sure that your contracts are well-documented,
  - run the Solidity linter (`npm run lint:sol`) and fixed any issues,
  - run the JS linter and fixed any issues (`npm run lint:fix`), and
  - updated the changelog, if applicable.
-->

frangio commented on contracts/presets/ERC20PresetFixedSupply.sol:
```diff
@@ -0,0 +1,21 @@
+        string memory symbol,
+        uint256 initialSupply
+    ) public ERC20(name, symbol) {
+        _mint(msg.sender, initialSupply);
```
I think it would be better to use an address in the constructor arguments instead of `msg.sender`. This improves "composability", for example it allows you to use a generic factory to deploy the contract and assign the initial balance to a separate account.

In hindsight this may be an issue with the current roles presets, since they assign roles to the deployer account.

ashwinYardi replied:
Got it! Replacing msg.sender by an argument 'owner' of type address.

frangio commented on test/presets/ERC777PresetFixedSupply.js:
```diff
@@ -0,0 +1,53 @@
+
+const ERC777PresetFixedSupply = artifacts.require('ERC777PresetFixedSupply');
+
+contract('ERC777', function (accounts) {
```
```suggestion
contract('ERC777PresetFixedSupply', function (accounts) {
```

frangio commented on contracts/presets/ERC20PresetFixedSupply.sol:
```diff
@@ -0,0 +1,21 @@
+ * Tip: Contract is preconfigured with fixed initial supply.
+ * minting / pausing functionality is not supported. This removes the need of access control and thereby governance. 
+ */
+contract ERC20PresetFixedSupply is ERC20, ERC20Burnable {
```
What was the reason for including `ERC20Burnable`?

ashwinYardi replied:
Burn functionality would enable the use of this preset for the purposes of deflationary tokens, correct?

Skyge replied:
Emmm, I am confused the can I just write `contract ERC20PresetFixedSupply is ERC20Burnable`, cause `ERC20Burnable` is derived from `ERC20`

ashwinYardi replied:
ahh, yes you are right! Thanks @Skyge

frangio commented on test/presets/ERC777PresetFixedSupply.js:
```diff
@@ -0,0 +1,53 @@
+
+  const defaultOperators = [defaultOperatorA, defaultOperatorB];
+
+  context('with default operators', function () {
```
You should be testing the balance of the initial account.

frangio reviewed (commented):
Thanks @ashwinYardi!

ashwinYardi commented:
@frangio @Skyge Some info about the testcases included: 

Since the base contracts used in above presets, already have a full fledged test cases to cover the entire behavior, Only bare minimum test cases which help in establishing the high level behavior of preset contracts have been added. Again, I referred to test cases written for already added presets and it made sense to proceed this way. 

Thanks and please let me know if there is anything missing or needs to be corrected, would be happy to do it immediately so that we can get this meged! ( PS. Sorry for not mentioning this info about testcases on PR's description ).

frangio reviewed (commented):
@ashwinYardi Sorry for taking so long to come back to this!

Thank you for explaining the rationale for what to test. I didn't remember the other presets followed the same approach to testing.

I'd like to see a few more changes in the tests before we merge this:

- Remove the `context` and `describe` "wrappers", since they don't serve any purpose here.
- Remove tests that we can assume correct from the base ERC777 contract:
    - returns a granularity of 1
    - returns 18 when decimals is called
- ERC20: add a test for the total supply

I also think the documentation needs a little bit of work. But this is something we can do if you prefer. If you'd like to take a stab at it you can take a look at the docs comments for the existing presets and follow a similar format. The docs for the constructor are missing too.

In order for the presets to show in the site, you need to add their names in `contracts/presets/README.adoc`.

ashwinYardi commented:
Hi @frangio ! Thanks a lot for the feedback. 

- [x] Removed the unused context and describe "wrappers"
- [x] Removed the suggested tests that can be assumed correct from base implementation
- [x] Added a test for total supply in ERC20 preset
- [x] Added a test for decrementing the total supply when tokens are burnt in ERC20 preset
- [x] Gave a shot at improving documentation
- [x] Added required entries in `contracts/presets/README.adoc`

frangio reviewed (approved):
Great work @ashwinYardi! I made a few small changes to the docs, added a changelog entry, and I realized I prefer to keep the `defaultOperators` argument in the same place as the `ERC777` constructor so I changed that as well.
